Update ecflow package for wcoss2 GFS transition

lgannoaa opened the pull request:
This PR:
- No need to review obsproc related files and link under ecflow. NCO will work on new ops version.
- Updates the 4 suite definition files prodXX.def; XX=00,06,12,18
- Includes updates to facilitate developer testing by specifying developer values for variables such as QUEUE, 
   QUEUE_TRANSFER, QUEUE_SHARED and COM
- Modify module position in each ecflow job using the same position from 
  $HOMEgfs/modulefiles/module_base.wcoss2.lua
- Adding envir-p1.h for wcoss2 environment
- Include head.h for reference developer setting
- Change in ecflow script:
  - Updates the ecflow scripts to consistent with GFS module position
  - Using wcoss2 environment
  - Using PBS for WCOSS2
  - Using updated module 
  - Move enkf out of gdas and rename it to enkfgdas.
     Include all ecflow definition files job name
     Include all ecflow scripts name and job/log name
  - Move "model=gfs" to the top on each job except all jobs under obsproc.
     obsproc will no longer be part of GFS. Therefore leave it without change for testing purpose.
  - Remove the source of model_ver from each ecflow script except all jobs under obsproc.
     obsproc will no longer be part of GFS. Therefore leave it without change for testing purpose.
This work is part of #398

lgannoaa committed:
ecflow package for wcoss2 GFS transition
  WCOSS2 Migration and Porting
    https://github.com/NOAA-EMC/global-workflow/issues/398

aerorahul commented on ecf/scripts/gdas/atmos/obsproc/dump/jgdas_atmos_dump_alert.ecf:
```diff
@@ -1,39 +1,64 @@
+export obsNET=obsproc
+
+#### Developer overwrite
+####export KEEPDATA="YES"
```
Please remove commented code.

aerorahul replied:
Per @lgannoaa This will be removed by NCO.

aerorahul commented on ecf/include/envir-p1.h:
```diff
@@ -0,0 +1,46 @@
+export MACHINE_SITE=%MACHINE_SITE%
+export SENDDBN=${SENDDBN:-%SENDDBN:YES%}
+export SENDDBN_NTC=${SENDDBN_NTC:-%SENDDBN_NTC:YES%}
+#module load prod_envir/2.0.4 prod_util/2.0.9 envvar/1.0
+echo debug
```
Please remove commented and unnecessary `echo`

lgannoaa replied:
done

aerorahul commented on ecf/scripts/gdas/atmos/gempak/jgdas_atmos_gempak.ecf:
```diff
@@ -43,6 +38,9 @@ export USE_CFP=YES
 ############################################################
 export model=gdas
+
+#export COMPATH=/lfs/h1/ops/canned/com/ukmet:/lfs/h1/ops/canned/com/gfs:/lfs/h1/ops/canned/com/ecmwf:/lfs/h1/ops/canned/com/nam:/lfs/h1/ops/canned/com/hourly
```
please remove commented line or put a comment that this is necessary for ops and this is a Developer Overwrite

aerorahul replied:
Per @lgannoaa NCO will remove this commented line.

aerorahul commented on ecf/scripts/gdas/atmos/init/jgdas_atmos_gldas.ecf:
```diff
@@ -43,6 +37,7 @@ module list
 #############################################################
 export cyc=%CYC%
 export cycle=t%CYC%z
+export CDATE=$PDY$cyc
```
This variable is set in `envir-p1.h`.  It does not need to be set here.

lgannoaa replied:
Removed

aerorahul commented on ecf/scripts/gfs/atmos/gempak/jgfs_atmos_gempak_ncdc_upapgif.ecf:
```diff
@@ -24,11 +21,11 @@ model=gfs
+module load gempak/${gempak_ver}
+module load libjpeg/${libjpeg_ver}
+module load grib_util/${grib_util_ver}
+module load imagemagick/${imagemagick_ver}
```
Where is `${imagemagic_ver}` coming from?  I don't see it [here](https://github.com/NOAA-EMC/global-workflow/blob/feature/ops-wcoss2/versions/run.ver)

lgannoaa replied:
This job need module imagemagic

aerorahul replied:
Yes. I understand that. But what is the value of `imagemagik_ver` and where is it set?

lgannoaa replied:
It need to be set in run.ver

aerorahul commented on ecf/scripts/gfs/atmos/post/jgfs_atmos_post_f000.ecf:
```diff
@@ -1,24 +1,16 @@
+#PBS -q %QUEUE%
+#PBS -A %PROJ%-%PROJENVIR%
+#PBS -l walltime=00:20:00
+#PBS -l place=vscatter,select=10
```
@WenMeng-NOAA On WCOSS_Dell, this job (and the other `jgfs_atmos_post_fXXX` jobs), they take 4 Nodes totaling to 112 tasks.  See lines 17,18 on the left.
Am I reading this right that this job (and other `jgfs_atmos_post_fXXX` jobs) are requesting 10 nodes totaling to 1280 tasks for a single forecast hour.

aerorahul commented on ecf/scripts/gfs/atmos/post/jgfs_atmos_post_manager.ecf:
```diff
@@ -1,24 +1,16 @@
+#PBS -S /bin/bash
+#PBS -N gfs_atmos_post_manager_%CYC%
+#PBS -j oe
+#PBS -q %QUEUE%
```
Surely this job can run in the shared queue as it only sleeps.

WalterKolczynski-NOAA replied:
Per Kate today, shared is the default

aerorahul commented on ecf/scripts/gfs/atmos/post/jgfs_atmos_post_manager.ecf:
```diff
@@ -1,24 +1,16 @@
+#PBS -q %QUEUE%
+#PBS -A %PROJ%-%PROJENVIR%
+#PBS -l walltime=04:00:00
+#PBS -l place=vscatter,select=1:ncpus=128
```
Surely this job only needs 1 task as it only sleeps

aerorahul reviewed (changes requested):
I have reviewed upto the file:
` ecf/scripts/gfs/atmos/post_processing/bufr_sounding/jgfs_atmos_postsnd.ecf`
But not `head.h` as it is very complicated.

Also PR #553 has been merged, so please rebase this PR based on current `feature/ops-wcoss2`

Please see comments.

WalterKolczynski-NOAA commented on ecf/versions/obsproc.ver:
```diff
@@ -0,0 +1,69 @@
+export dictionaries_ver="v3.4.0"
+export omb_ver="v4.4"
+
+## main "model" software
```
All these commented  out lines should be removed.

WalterKolczynski-NOAA commented on ecf/versions/gfs.ver:
```diff
@@ -1,30 +1,60 @@
+export radarl2_ver=v1.2
+
+#### NCO requested testing location is v16.2
+#### export gfs_ver="v16.2.0"
```
Removed unused line

WalterKolczynski-NOAA reviewed (changes requested):
Reviewed from jgfs_atmos_post_anl.ecf to the end.

aerorahul commented on ecf/include/envir-p1.h:
```diff
@@ -0,0 +1,46 @@
+    else 
+      export SIPHONROOT=/lfs/h1/ops/prod/dbnet_siphon
+    fi
+#### Developer overwrite
+    export SIPHONROOT=${UTILROOT}/fakedbn
+####
+else
+    export SIPHONROOT=${UTILROOT}/fakedbn
+fi
```
This is developer only.  Not used by NCO and thus will not be addressed by NCO.
Please resolve the if-else logic as it is executed regardless of the condition.

lgannoaa replied:
Removed

aerorahul commented on ecf/include/envir-p1.h:
```diff
@@ -0,0 +1,46 @@
+
+export DBNROOT=$SIPHONROOT
+
+####if [[ ! " prod para test " =~ " ${envir} " && " ops.prod ops.para " =~ " $(whoami) " ]]; then err_exit "ENVIR must be prod, para, or test [envir-p1.h]"; fi
```
Why is this commented out? If it is unused, what purpose does it serve?
Either put a comment on this explaining the purpose or remove.

lgannoaa replied:
Restored

aerorahul commented on ecf/include/head.h:
```diff
@@ -4,45 +4,72 @@ export PS4='+ $SECONDS + '
-    module load EnvVars/$EnvVars_ver
-    module load ecflow/$ecflow_ver
-    POST_OUT=/var/lsf/ecflow_post_in.$USER.$LSB_BATCH_JID
+export ECF_RID=${ECF_RID:-${PBS_JOBID:-$$}}
```
Per @lgannoaa NCO does not use any files from under the `include` directory.
These are for developer use only.

aerorahul commented on ecf/include/envir-p1.h:
```diff
@@ -0,0 +1,46 @@
+
+####if [[ ! " prod para test " =~ " ${envir} " && " ops.prod ops.para " =~ " $(whoami) " ]]; then err_exit "ENVIR must be prod, para, or test [envir-p1.h]"; fi
+export ECF_PORT=34326
+export DATAROOT=/lfs/h2/emc/stmp/Lin.Gan/RUNDIRS/ecfops
+export COMROOT=/lfs/h2/emc/ptmp/Lin.Gan/ecfops/com
```
Please replace `Lin.Gan` with `$USER`.

lgannoaa replied:
Done

aerorahul replied:
I still see `Lin.Gan` instead of `$USER`

aerorahul commented on ecf/include/envir-p1.h:
```diff
@@ -0,0 +1,46 @@
+export DBNROOT=$SIPHONROOT
+
+####if [[ ! " prod para test " =~ " ${envir} " && " ops.prod ops.para " =~ " $(whoami) " ]]; then err_exit "ENVIR must be prod, para, or test [envir-p1.h]"; fi
+export ECF_PORT=34326
```
Please remove this line.  This variable is set when `ecflow` module is loaded.  `ecflow` module is loaded in `head.h`. This variable is unique to the user.

lgannoaa replied:
The job tested without this line sometime failed silently. There is a good reason to include this in here.
As discussed, this file will be used only for reference. I prefer to keep it.

aerorahul replied:
Sometimes failure is a symptom of the system instability.  I would rather encounter this failure sometimes, than hard-wire some users personal ECF_PORT information.  I suggest removal.

lgannoaa replied:
Removed

aerorahul commented:
This PR has been partially reviewed.  
The changes are too many to be made at this time for a thorough cleanup and review and retest.
Please merge as it only affects ecflow and only a single developer can run this setup in its current form due to hardwired paths, username, and `ECF_PORT` number in the ecflow files such as `.def`, `head.h`, `model_ver.h`, `envir_p1.h`.

aerorahul commented on ecf/scripts/gdas/atmos/analysis/jgdas_atmos_analysis.ecf:
```diff
@@ -29,21 +21,21 @@ export model=gfs
-export KMP_AFFINITY=scatter
 export OMP_NUM_THREADS=$threads
 export FORT_BUFFERED=true
 export OMP_NUM_THREADS_CY=28
```
Is `OMP_NUM_THREADS_CY=28` appropriate for WCOSS2?  `config.resources.nco.static` is setting this to `npe_node_max`. `28` is the value for `WCOSS_Dell_P3` and was hard-wired here.

aerorahul replied:
The same is the case in `jgfs_atmos_analysis.ecf`.

lgannoaa replied:
Set to 128

aerorahul replied:
> Set to 128

Are you sure, the value from this variable set here `OMP_NUM_THREADS_CY` is being used?
I see https://github.com/NOAA-EMC/GSI/blob/a9561f4219e5a69ac1af250081ae027084c40c3b/scripts/exglobal_atmos_analysis.sh#L1017 setting the value based on `NTHREADS_CYCLE` right before entering `CYCLESH`

aerorahul commented on ecf/scripts/gdas/atmos/analysis/jgdas_atmos_analysis.ecf:
```diff
@@ -29,21 +21,21 @@ export model=gfs
 # WCOSS environment settings
 #############################################################
-export KMP_AFFINITY=scatter
 export OMP_NUM_THREADS=$threads
```
Where is the value for `$threads` coming from?  It was previously set on line 20 above and has been removed in this PR.

lgannoaa replied:
Removed.

lgannoaa committed:
Merge remote-tracking branch 'upstream/feature/ops-wcoss2' into feature/ecflow-ops-ecf2

lgannoaa committed:
Update enkf structure changes in ecflow definition files

lgannoaa committed:
Reference to NCO version:
 - Move enkf out of gdas and rename it to enkfgdas.
     Include all ecflow definition files job name
     Include all ecflow scripts name and job/log name
 - Move "model=gfs" to the top on each job except all jobs under obsproc.
     obsproc will no longer be part of GFS. Therefore leave it without change for testing purpose.
 - Remove the source of model_ver from each ecflow script except all jobs under obsproc.
     obsproc will no longer be part of GFS. Therefore leave it without change for testing purpose.

lgannoaa committed:
Remove remark from envir-p1.h and head.h
Update analysis ecflow script to use 128 for wcoss2
Remove extra CDATE

lgannoaa committed:
Remove ecflow post assignment in envir-p1.h

aerorahul commented on ecf/include/envir-p1.h:
```diff
@@ -0,0 +1,41 @@
+export MACHINE_SITE=%MACHINE_SITE%
+export SENDDBN=${SENDDBN:-%SENDDBN:YES%}
+export SENDDBN_NTC=${SENDDBN_NTC:-%SENDDBN_NTC:YES%}
+if [[ "$envir" == prod && "$SENDDBN" == YES ]]; then
+    export eval=%EVAL:NO%
+    if [ $eval == YES ]; then 
+      export SIPHONROOT=${UTILROOT}/para_dbn
+    else 
+      export SIPHONROOT=/lfs/h1/ops/prod/dbnet_siphon
+    fi
+    export SIPHONROOT=${UTILROOT}/fakedbn
+else
+    export SIPHONROOT=${UTILROOT}/fakedbn
+fi
```
Based on the logic here, regardless of the conditions, the result is:
`export SIPHONROOT=${UTILROOT}/fakedbn`
Also, a variable `eval` is being exported for use somewhere down when `envir` == `prod` and `SENDDBN` is `YES`.
Is that the desired outcome?

lgannoaa replied:
The developer will always use fakedbn to test. eval is ok here

aerorahul replied:
What does `eval` do?  Is it used anywhere other than the if-test?  
If the developer will always use fakeddbn, then the first if test `"$envir" == prod` is not necessary (and this whole block just boiled down to this):
```
export SIPHONROOT=${UTILROOT}/fakedbn
```

aerorahul commented on ecf/defs/prod00.def:
```diff
@@ -1,6 +1,6 @@
 extern /prod18/gdas/jgdas_forecast
 extern /prod18/gdas/atmos/post
-extern /prod18/gdas/enkf/post
+extern /prod18/enkfgdas/post
```
These are new changes added in these `.def` files that were agreed to be added in a follow-up PR in the meeting this AM.  This PR is already big to add any additional changes.

aerorahul commented on ecf/include/head.h:
```diff
@@ -4,45 +4,68 @@ export PS4='+ $SECONDS + '
+  module load ecflow/$ecflow_ver
+  echo "ecflow module location: $(module display ecflow |& head -2 | tail -1 | sed 's/:$//')"
+  export ECF_ROOT=/apps/ops/prod/nco/core/ecflow.v5.6.0.7
+  export ECF_PORT=34326
```
Please remove user specific PORT information.  This value is set in the `ecflow` module for the user and is unique to the user.

aerorahul commented on ecf/include/head.h:
```diff
@@ -4,45 +4,68 @@ export PS4='+ $SECONDS + '
+  echo "Running module load ecflow/$ecflow_ver"
+  module load ecflow/$ecflow_ver
+  echo "ecflow module location: $(module display ecflow |& head -2 | tail -1 | sed 's/:$//')"
+  export ECF_ROOT=/apps/ops/prod/nco/core/ecflow.v5.6.0.7
```
There are dual uses of the variable `ecflow_ver` and `5.6.0.7` for ecflow version.  Which one is it?  Please be consistent in the use of the version.

aerorahul replied:
This variable is set by the `ecflow` module.  It should not be overwritten.  Load the appropriate module instead.
Please remove.

aerorahul commented on ecf/include/head.h:
```diff
@@ -4,45 +4,68 @@ export PS4='+ $SECONDS + '
+export gfs_ver=v16.2
+
+if [ -d /apps/ops/prod ]; then # On WCOSS2
+  export ECF_ROOT=/apps/ops/prod/nco/core/ecflow.v5.6.0.7
+  . ${ECF_ROOT}/versions/run.ver
```
I am failing to understand what this is doing.  Why `export`? and why loading `prod_util` in `head.h`

aerorahul commented on ecf/include/head.h:
```diff
@@ -4,45 +4,68 @@ export PS4='+ $SECONDS + '
+  export ECF_ROOT=/apps/ops/prod/nco/core/ecflow.v5.6.0.7
+  export ECF_PORT=34326
+  export ECF_HOST=ddecflow02
+  export ECF_INCLUDE=/lfs/h2/emc/global/noscrub/$USER/git/feature-ops-wcoss2/ecf/include
```
This is hardwired to a specific branch which will not exist as soon as it is merged and deleted.
I am failing to understand the need for this entire section from 40-47.  It seems to be tailored for a specific branch and specific location.

aerorahul commented on ecf/scripts/enkfgdas/analysis/create/jenkfgdas_diag.ecf:
```diff
@@ -0,0 +1,51 @@
+#PBS -S /bin/bash
```
These EnKF jobs are new and we agreed to add them in a followup PR to avoid further bloating this current PR.

aerorahul reviewed (changes requested):
Please see more comments.

KateFriedman-NOAA commented:
Forcibly submitting this PR due to need.
